src/lightning/data/cache/dataset_optimizer.py

```diff
@@ -0,0 +1,761 @@
+            end_node = len(user_items) if is_last_node else (node_rank + 1) * node_size
+            node_user_items = user_items[start_node:end_node]
+            worker_size = len(node_user_items) // self.num_workers
+            for worker_idx in range(self.num_workers):
+                is_last = worker_idx == self.num_workers - 1
+                begin = worker_idx * worker_size
+                end = len(node_user_items) if is_last else (worker_idx + 1) * worker_size
+                workers_user_items.append(user_items[begin:end])
+                begins.append(begin)
+            return begins, workers_user_items
+        raise RuntimeError(f"The current_node_rank {current_node_rank} doesn't exist in {num_nodes}.")
+
+    def _cached_list_filepaths(self) -> List[str]:
+        """This method lists and caches the."""
+        home = _get_home_folder()
+        filepath = os.path.join(home, ".cache", f"{self.name}/filepaths.txt")
+
+        if os.path.exists(filepath):
+            lines = []
+            with open(filepath) as f:
+                for line in f.readlines():
+                    lines.append(line.replace("\n", ""))
+            return lines
+
+        str(Path(self.src_dir).resolve())
+
+        filepaths = []
+        for dirpath, _, filenames in os.walk(self.src_dir):
+            for filename in filenames:
+                filepaths.append(os.path.join(dirpath, filename))
+
+        os.makedirs(os.path.dirname(filepath), exist_ok=True)
+
+        with open(filepath, "w") as f:
+            for filepath in filepaths:
+                f.write(f"{filepath}\n")
+
+        return filepaths
+
+    def _signal_handler(self, signal: Any, frame: Any) -> None:
+        """On temrination, we stop all the processes to avoid leaking RAM."""
+        for w in self.workers:
+            w.join(0)
+        os._exit(0)
+
+    def _upload_index(self, cache_dir: str, num_nodes: int, node_rank: Optional[int]) -> None:
+        """This method upload the index file to the remote cloud directory."""
+        if not self.remote_dst_dir:
+            return
+
+        obj = parse.urlparse(self.remote_dst_dir)
+        if num_nodes > 1:
+            local_filepath = os.path.join(cache_dir, f"{node_rank}-{_INDEX_FILENAME}")
+        else:
+            local_filepath = os.path.join(cache_dir, _INDEX_FILENAME)
+
+        if obj.scheme == "s3":
+            s3 = boto3.client("s3")
+            s3.upload_file(
+                local_filepath, obj.netloc, os.path.join(obj.path.lstrip("/"), os.path.basename(local_filepath))
+            )
+        elif os.path.isdir(self.remote_dst_dir):
+            copyfile(local_filepath, os.path.join(self.remote_dst_dir, os.path.basename(local_filepath)))
+
+        if num_nodes == 1 or node_rank is None:
+            return
+
+        # Merge the index files generated by each node
+        if num_nodes == node_rank + 1:
+            # Get the index file locally
+            for node_rank in range(num_nodes - 1):
```

Comment:
Which processes are executing this? I don't see any guards, so do all ranks run this? It's unclear whether there will be race conditions

Reply:
This isn't linked to PyTorch Lightning. There is a single process per node, so no race condition possible there.

Reply:
Added a comment.